### ISSUE TITLE
Research additional predictive modeling techniques

### ISSUE THREAD

**Issue by nicklucius** [Chicago/e-coli-beach-predictions#109]
Look into enhancements for our data preprocessing and modeling.

For example, we could try [downsampling](https://www.r-bloggers.com/down-sampling-using-random-forests/) as a way to deal with an unbalanced response variable, (i.e., we have way more high-level e. coli days than low-level.)

![image](https://cloud.githubusercontent.com/assets/16853555/20403663/4e8c9c90-acc7-11e6-8847-e4be5505a571.png)

**Comment by nicklucius** [comment]
Now that the code is reworked and the new dataset is nearly ready (#107 and #108), I wanted to record the thoughts that @CallinOsborn and I have talked about, many of which were discussed at Hack Night meetings earlier this year. I hope these ideas provide a springboard for new models.

## Redo models with new water level data and new engineered features

Many of us believe elevated E. coli levels are the result of a complex system, where levels jump up if 
all the right conditions are present. For instance, one pattern of E. coli contamination could start 
with a period of no rain and good weather. During this time, the beaches accumulate with humans and animals, and the resulting fecal matter. If the water levels are low, then the beach area is larger, 
which allows more accumulation of feces. Remember that E. coli lives outside a body only for a limited time period. When the E. coli is still alive, a large rainstorm and/or sudden rise in water levels could wash it all out into the lake. Right after the "wash", E. coli levels would be high until the E. coli naturally dies or dissipates.

And then there's the fact that direct sunlight can kill E. Coli. 

That may or may not describe an actual series of events that leads to high E. coli levels at one or more beaches. And other complex systems could exist. But if that is an accurate description, our model would need certain engineered predictors to catch the pattern and provide a good prediction. We could engineer features such as:
- Days since last rain
- Days since temp went below X

And so on. The point is, it is possible that the data contains more answers than our current model can find without more engineered features. These new features may be the key to improving the model's performance.

### Wind Speed and Direction

With the addition of beach-facing direction, we could benefit from adding the wind direction. The unit measure of `windBearing` is degrees. This poses a problem since 1 degree and 359 degrees are very close in reality, but not in numerical representation. We could engineer a feature than helps interpret the wind in relation to the beach-facing direction.

## Evaluate Global Model vs. Beach Models

So far, our models have been single, global models (and ensembles) that use the same formula to attempt to predict the E. coli level at each beach. We have controlled for beach within the models, but we could try evaluating a modelling technique that creates an individual model for each beach. Also, the global models themselves could be improved by using beach characteristics as predictors, such  as 
- Beach Face Direction (e.g. North, East)
- Beach Area
- Breakwater characteristics

## Study the effects of precipitation

Many different analyses of the data have pointed to `precipIntensity` being a major predictor. While 
modelling has confirmed its predictive power, there is a limit to its effectiveness. Perhaps it is 
because the right conditions must be present before rain can wash E. coli into the Lake. It might be worth studying high E. coli days that occur right after a big rain. Isolating those occurrences and 
making data visualizations could help us understand the underlying conditions that accompany high-levels of E. coli. Once we understand better, we can engineer features that capture the information in order to pass it to a model.

**Comment by nicklucius** [comment]
I looked at the relationship between same-day E. Coli levels at different beaches. Here are some exploratory visualizations to talk about.

![image](https://cloud.githubusercontent.com/assets/16853555/20674133/56b1ed1a-b54d-11e6-8453-ec4ac0c0d733.png)

![image](https://cloud.githubusercontent.com/assets/16853555/20674158/6cad20ee-b54d-11e6-90f6-3e280b81dde5.png)

![image](https://cloud.githubusercontent.com/assets/16853555/20674180/823e47d0-b54d-11e6-9761-743d9a5c6236.png)

**Comment by nicklucius** [comment]
![image](https://cloud.githubusercontent.com/assets/16853555/20674537/fcbc9056-b54e-11e6-983a-3603483429ff.png)

**Comment by nicklucius** [comment]
As we all know, we have been trying to predict today’s E. coli level by using yesterday’s (and older) data as predictors. After many tries, we have not been able to produce a model that does better than the current USGS model on new observations.

Recently, I’ve looked into a different technique for predicting today’s E. coli level by taking today’s DNA test results at a sample of beaches, and then using those readings as predictors to forecast the E. coli level at the other beaches that were not DNA-tested. Initial results show that same-day tests may have the power to increase the accuracy of E. coli predictions by many times over the USGS model.

## USGS Performance

First, to establish a baseline for performance, below are the confusion matrices to show the USGS model’s track record so far. We have data on 2015 and 2016.

2015 USGS Model (all beaches)

| Precision: 48% | Predicted High | Predicted Normal |
| --- | --- | --- |
| Actual High | 12 (6.5%) | 170 (92.5%) |
| Actual Normal | 13 (1.1%) | 1207 (98.9%) |

2016 USGS Model (all beaches)

| Precision: 45% | Predicted High | Predicted Normal |
| --- | --- | --- |
| Actual High | 14 (10.6%) | 118 (89.4%) | 
| Actual Normal | 17 (1.5%) | 1106 (98.5%) | 

As you can see, the model’s true positive rate ranged between 6% and 11% and precision ranged between 45% and 48%.

## Prior-Day Model Performance

The models our team created were developed between the 2015 and 2016 swim seasons, so for comparison I’ll look at our performance on 2016 only. And for simplicity, I will detail one model which had the highest performance in 2016 of any of our models. As a caveat, I re-tuned the model after the 2016 beach season, so these results are slightly overconfident. The most significant difference is that I added a downsampling component, which increased performance, but not enough to match the USGS model.

2016 Prior-Day Model (all beaches)

| Precision: 39% | Predicted High | Predicted Normal |
| --- | --- | --- |
| Actual High | 12 (8.8%) | 124 (91.2%) | 
| Actual Normal | 19 (1.6%) | 1131 (98.4%) | 

To compare our model with the USGS model, I configured it to achieve a similar false positive rate at around 1.5%. At this configuration, our best model slightly underperformed the true positive rate and precision of the USGS model. There were no configurations at which our model significantly beat the USGS model.

See the end of this report for more details on the models described here.

## Same-Day Test Analysis

DNA tests, which produce results a few hours after sample is taken, provided me with a new way to think about this prediction problem. In 2015 and 2016, 5 beaches were tested using the DNA method, and we have the results for 2016 in the data model already. Given the same-day E. coli results at 5 of the 20 beaches, how well can we guess the levels at other 15 other beaches? In a way, it is no longer a future prediction problem. If you think about the 20 beaches as a population, we are trying to guess the population, given a sample. Perhaps there are inferential statistical methodologies that could improve the model’s performance. But to accomplish a quick study, I continued with our predictive modeling framework.

I made the decision to add Howard to the list of beaches for which we would have same-day DNA test results, even though it is not currently DNA tested. There is a cluster of beaches on the North side that usually have a similar E. coli level, and Howard is normally representative of those beaches. I assumed that having at least one beach from this cluster would be important.

## DNA vs. Culture tests

There are important differences between DNA test results and culture test results. The units of measurement are different, and city officials use different thresholds for issuing advisories. 

Also, we have DNA data on the Data Portal for 2016 only, and we have a CSV containing 2015 DNA data. For this analysis, I have only used the 2016 data as the CSV has not been imported to our R framework yet. 

Since we do not have a long history of DNA data, I decided to create two models to test the idea of sampling beaches with same-day DNA tests. 

- **Geomean Model**: Since we have 10 years of culture test geomeans for all the beaches, I created a model to determine how well the geomeans at 6 beaches can predict the geomeans at the remaining 14 beaches.

- **DNA-mean Model**: I created a second model to determine how well the DNA tests at 5 beaches (and the geomean at Howard) can predict the culture geomeans at the remaining 14 beaches. I split 2016 data into 2 randomly assigned train/test sets.

The Geomean Model showed that by including same-day culture test results at 6 beaches, we can vastly improve on the USGS model’s performance on the remaining 14 beaches. In all the curves that follow, the blue line is the model being evaluated, and the red line is the USGS model.

![image](https://cloud.githubusercontent.com/assets/16853555/21064823/be7c2d10-be22-11e6-8d1c-d858371f501d.png)

![image](https://cloud.githubusercontent.com/assets/16853555/21064849/e2557fd4-be22-11e6-959a-7459e81a3291.png)

2015-2016 USGS Model (14 beaches)

| Precision: 18% | Predicted High | Predicted Normal |
| --- | --- | --- |
| Actual High | 3 (2.2%) | 131 (97.8%) | 
| Actual Normal | 14 (0.8%) | 1606 (99.2%) | 

2015-2016 Geomean Model (14 beaches)

| Precision: 47% | Predicted High | Predicted Normal |
| --- | --- | --- |
| Actual High | 21 (15.6%) | 113 (84.4%) | 
| Actual Normal | 24 (1.5%) | 1596 (98.5%) | 

To compare our model with the USGS model, I again configured it to achieve a similar false positive rate that we used above (1.5%). At this configuration, our best model significantly improved the true positive rate and precision over the USGS model. 

And, just a small increase in the false positive rate resulted in a doubling of the true positive rate, from 15.6% to 32% at a slightly lower precision rate of 40%.

2015-2016 geomean model (14 beaches)

| Precision: 40% | Predicted High | Predicted Normal |
| --- | --- | --- |
| Actual High | 43 (32.0%) | 91 (68.0%) | 
| Actual Normal | 65 (4.0%) | 1555 (96.0%) | 

But there is a problem with predicting same-day E. coli levels using same-day culture tests—they do not exist. We have to determine whether same-day DNA tests are as predictive as same-day culture tests.
We do not have 10 years of DNA test results. 

So, to determine whether DNA test readings can match the performance of geomean readings as predictors, I did a 50/50 split of 2016 DNA test data into training and testing sets. I made sure that there was no “day bleeding”: I used ```Day_of_year``` to assign odd numbers to one set and even numbers to the other.

Keeping in mind this is a small dataset, the results show that the DNA-mean does give a significant boost to the model's performance.

![image](https://cloud.githubusercontent.com/assets/16853555/21064975/989ffdaa-be23-11e6-92d9-c324c75842c3.png)

![image](https://cloud.githubusercontent.com/assets/16853555/21064977/9b1e6d82-be23-11e6-89db-d881499e7743.png)

2016 USGS model (14 beaches)

| Precision: 44% | Predicted High | Predicted Normal |
| --- | --- | --- |
| Actual High | 4 (9.0%) | 40 (91.0%) | 
| Actual Normal | 5 (1.7%) | 287 (98.3%) | 

2016 DNA-mean model (14 beaches)

| Precision: 60% | Predicted High | Predicted Normal |
| --- | --- | --- |
| Actual High | 6 (13.6%) | 38 (86.4%) | 
| Actual Normal | 4 (1.3%) | 288 (98.7%) | 

As above, these two confusion matrices are configured at about the same false positive rate. Below, you can see the effect of allowing a few more false positives.

2016 DNA-mean model (14 beaches)

| Precision: 55% | Predicted High | Predicted Normal |
| --- | --- | --- |
| Actual High | 16 (36.3%) | 28 (63.7%) | 
| Actual Normal | 13 (4.4%) | 279 (95.6%) | 

## Model Details:

**2016 Prior-Day Model (all beaches)**

- Type: Random Forest
- Predictors
   - ```precipProbability```
   - ```1.daysPrior.precipIntensity```
   - ```Client.ID```
- Downsampling
   - 50/50 split between “low” days and “high” days
   - Low – under 75 CFU
   - High – 235 or higher CFU
- Training set: 2006 – 2015, all beaches, NA handling = use only rows with no NAs
- Validation type: Leave-one-year-out
- Final holdout set: 2016, all beaches, NA handling = use only rows with no NAs

**Geomean Model**
- Type: Random Forest
- Predictors
   - ```1.days.prior.precipIntensity```
   - ```precipProbability```
   - ```63rd_geomean```
   - ```South Shore_geomean```
   - ```Montrose_geomean```
   - ```Calumet_geomean```
   - ```Rainbow_geomean```
   - ```Howard_geomean```
   - ```Client.ID```
- Downsampling: NONE
- Training set: 2006 – 2014, 14 beaches, NA handling = use only rows with no NAs
- Validation type: Leave-one-year-out
- Final holdout set – 2015-2016, 14 beaches, NA handling = use only rows with no NAs

**DNA-mean Model**
- Type: Random Forest
- Predictors
   - ```1.days.prior.precipIntensity```
   - ```precipProbability```
   - ```63rd_DNAmean```
   - ```South Shore_DNAmean```
   - ```Montrose_DNAmean```
   - ```Calumet_DNAmean```
   - ```Rainbow_DNAmean```
   - ```Howard_geomean``` (no DNA mean available)
   - ```Client.ID```
- Downsampling: NONE
- Training set: 2016 50/50 random split, 14 beaches, NA handling = use only rows with no NAs
  - No DNAmean prior to 2016
  - Guaranteed that no days in training set are in final holdout set
- Validation type: final holdout only
- Final holdout set: 2016 50/50 random split, 14 beaches, NA handling = use only rows with no NAs

**Comment by kbrose** [comment]
@nicklucius Amazing work! Have you tried using a much simpler model to transform DNA to E. Coli levels? A random forest seems like it might be over-kill. If a simpler model (logistic/linear etc.) can achieve similar results that'd be even more encouraging.

**Comment by nicklucius** [comment]
@kbrose Thanks! Good point about random forest. It does seem that the modeling we're doing now would lend itself to something simpler. I've tried a basic linear regression, which does a lot better than USGS but not as good as random forest. I have wondered if it's the beach-to-beach differences that the random forest is picking up.

I've thought about trying a [multilevel model](http://www.stat.columbia.edu/~gelman/arm/). Let me know if you have any thoughts.

**Comment by kbrose** [comment]
Multilevel models seem like a good fit. If possible, I'd try both linear and logistic multilevel regressions.
